alf/utils/video_recorder.py

```diff
@@ -0,0 +1,323 @@
+    distribution, the probabilities are directly plotted. For a continuous
+    distribution, the curve will be approximated by sampled points for each
+    action dimension.
+    * action: plotted as heatmaps. Long type actions will be first
+    converted to one-hot encodings.
+
+    ``env_frame`` is a ``numpy.ndarray`` with shape ``(x, y, 3)``,
+    representing RGB values for an x-by-y pixel image, output from
+    ``env.render('rgb_array')``.
+    """
+
+    def __init__(self,
+                 env,
+                 img_plot_width=640,
+                 value_range=1.,
+                 frames_per_sec=30,
+                 **kwargs):
+        """
+        Args:
+            env (Gym.env):
+            img_plot_width (int): the image width for displaying prediction info,
+                excluding the env frame width
+            value_range (float): if quantities plotted as heatmaps, the values
+                will be clipped according to ``[-value_range, value_range]``.
+            frames_per_sec (fps):
+        """
+        super(VideoRecorder, self).__init__(env=env, **kwargs)
+        self._img_plot_width = img_plot_width
+        self._dist_imgs_per_row = 4
+        self._value_range = value_range
+        self.frames_per_sec = frames_per_sec  # overwrite the base class
+
+    def capture_frame(self, pred_info=None):
+        """Render ``self.env`` and add the resulting frame to the video. Also
+        plot information in ``pred_info``.
+
+        Args:
+            pred_info (nested): a nest
+        """
+        if not self.functional: return
+        logger.debug('Capturing video frame: path=%s', self.path)
+
+        if pred_info is not None:
+            assert not self.ansi_mode, "Only supports rgb_array mode!"
+            render_mode = 'rgb_array'
+        else:
+            render_mode = 'ansi' if self.ansi_mode else 'rgb_array'
+
+        frame = self.env.render(mode=render_mode)
+
+        if frame is None:
+            if self._async:
+                return
+            else:
+                # Indicates a bug in the environment: don't want to raise
+                # an error here.
+                logger.warn(
+                    'Env returned None on render(). Disabling further '
+                    'rendering for video recorder by marking as disabled: '
+                    'path=%s metadata_path=%s', self.path, self.metadata_path)
+                self.broken = True
+        else:
+            if pred_info is not None:
+                frame = self._plot_pred_info(frame, pred_info)
+
+            self.last_frame = frame
+            if self.ansi_mode:
+                self._encode_ansi_frame(frame)
+            else:
+                self._encode_image_frame(frame)
+
+    def _plot_prob_curve(self, name, probs, xticks=None):
+        if xticks is None:
+            xticks = range(len(probs))
+
+        fig, ax = plt.subplots(figsize=(2, 2))
+        ax.plot(xticks, probs)
```

Comment:
The video on the plot curve part could to be more smooth if the frames are aligned, for example, by using the same axis range for all the plots in the same video/episode.

Reply:
It's difficult to determine the axis range beforehand. It's a density distribution so there is no upper bound on y.